default/python/universe_generation/empires.py

```diff
@@ -95,26 +85,75 @@ def min_planets_in_vicinity_limit(num_systems):
-    accepted_list = []
-    while attempts:
-        accepted = []
-        # get candidates that are at least min_jumps away from the systems we already accepted
-        for candidate in systems_min_jumps_away_from(systems_pool, min_jumps, accepted):
-            accepted.append(candidate)  # add the candidate to our list of accepted systems
-        # add the result of the current attempt to our list and continue with next attempt
-        accepted_list.append(accepted)
-        attempts -= 1
-    # return the result of the attempt that yielded the most systems
-    return max(accepted_list, key=len)
+class HomeSystemFinder(object):
+    """Finds a set of home systems with a least ''num_home_systems'' systems."""
+    def __init__(self, _num_home_systems):
+        # cache of sytem merits
+        self.system_merit = {}
+        self.num_home_systems = _num_home_systems
+
+    def find_home_systems_for_min_jump_distance(self, systems_pool, min_jumps):
+        """
+        Return a good list of home systems
+
+        Return a list of home systems which are at least the specified minimum number of jumps apart,
+        and with a good system merit picked randomly from the specified pool.
+
+        Return an empty list if there are fewer than num_home_systems in the pool.
+        """
+        # Make several attempts to find systems that match the condition
+        # of being at least min_jumps apart
+        # Use the minimum merit of the best num_home_system systems found
+        # as the merit of the current best set of systems.
+        # On each attempt use the merit of the current best set of home
+        # systems to truncate the pool of candidates.
+
+        # precalculate the system merits
+        for system in systems_pool:
+            if system not in self.system_merit:
+                self.system_merit[system] = calculate_home_system_merit(system)
+
+        best_case_worst_merit = sorted(self.system_merit.values())[self.num_home_systems - 1]
+        current_worst_merit = 0
+
+        best_candidate = []
+
+        # cap the number of attempts to the smaller of the number of systems in the pool, or 100
+        attempts = min(100, len(systems_pool))
+
+        while attempts:
+            candidate = []
+            # use a local pool of all candidate systems better than the worst threshold merit
+            local_pool = {s for s in systems_pool if self.system_merit[s] > current_worst_merit}
+            if len(local_pool) < self.num_home_systems:
+                break
+
+            while local_pool:
+                member = random.choice(list(local_pool))
+                candidate.append(member)
+
+                # remove all neighbors from the local pool
+                for neighbor in get_systems_within_jumps(member, min_jumps):
+                    if neighbor in local_pool:
+                        local_pool.remove(neighbor)
```

Comment:
Use python's set-functions for that:
```local_pool -= set(get_systems_within_jumps(member, min_jumps))``` or
```local_pool.difference_update(get_systems_within_jumps(member, min_jumps))```

https://docs.python.org/2/library/sets.html